Gosu/Window.h

```diff
@@ -7,16 +7,32 @@
     {
     struct Callbacks
     {
-      void (*update)();
-      void (*draw)();
+      void (*update)() = nullptr;
```

Comment:
Why do you store them in a nested `struct`?

I also think these callbacks should generally take a `Gosu_Window*` as their first arguments, so that the FFI-generated callbacks can pass it on to the right instance. Or does FFI generate a new stub for every object?

Reply:
Added a `void* data` pointer for callbacks to pass when called (Makes Crystal's Box work) https://github.com/gosu/gosu/pull/501/commits/77a6d5aa93cf0302b4ec04707c3cffc729d77022

How should I store the callbacks? In a struct implemented in the proper place or some other method?

Reply:
One idiomatic way in modern C++ would be to use std::function and lambdas to bind the callback & data pointer into a single object:

```cpp
#include <functional>
...
class WindowForWrapper : public Gosu::Window
{
public:
    void button_down(Gosu::Button btn) override
    {
        if (update_callback) update_callback(btn.id());
    }

    // Make these public to get rid of many internal setters?
    std::function<void (unsigned btn)> button_down_callback;
    ...
};

void Gosu_Window_set_button_down(Gosu_Window *window, void function(void *data, unsigned btn), void* data)
{
    reinterpret_cast<Gosu::WindowForWrapper*>( window )->button_down_callback = [=](unsigned btn) { function(data, btn); }
}
```

To be honest, your current solution is probably slightly more efficient than wrapping it in a `std::function`. You can also keep using raw function and void pointers, but I think getting rid of the internal setters would shorten the code somewhat.